substratevm/src/com.oracle.svm.jfr/src/com/oracle/svm/jfr/traceid/JfrTraceId.java

```diff
@@ -0,0 +1,158 @@
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package com.oracle.svm.jfr.traceid;
+
+import com.oracle.svm.core.annotate.Uninterruptible;
+import jdk.jfr.internal.JVM;
+import org.graalvm.compiler.api.replacements.Fold;
+import org.graalvm.nativeimage.ImageSingletons;
+import org.graalvm.nativeimage.Platform;
+import org.graalvm.nativeimage.Platforms;
+
+public class JfrTraceId {
+
+    private static final long MaxJfrEventId = 403;
+
+    public static final long BIT = 1;
+    public static final long META_SHIFT = 8;
+    public static final long TRANSIENT_META_BIT = (BIT << 3);
+    public static final long TRANSIENT_BIT = (TRANSIENT_META_BIT << META_SHIFT);
+    public static final long SERIALIZED_META_BIT = (BIT << 4);
+    public static final long SERIALIZED_BIT = (SERIALIZED_META_BIT << META_SHIFT);
+
+    private static final int TRACE_ID_SHIFT = 16;
+
+    // Epoch stuff
+    private static final long USED_BIT = 1;
+    private static final int EPOCH_1_SHIFT = 0;
+    private static final int EPOCH_2_SHIFT = 1;
+    private static final long USED_EPOCH_1_BIT = USED_BIT << EPOCH_1_SHIFT;
+    private static final long USED_EPOCH_2_BIT = USED_BIT << EPOCH_2_SHIFT;
+
+    private static final long JDK_JFR_EVENT_SUBKLASS = 16;
+    private static final long JDK_JFR_EVENT_KLASS = 32;
+    private static final long EVENT_HOST_KLASS = 64;
+
+    @Fold
+    static JfrTraceIdMap getTraceIdMap() {
+        return ImageSingletons.lookup(JfrTraceIdMap.class);
+    }
+
+    public static void tag(Class<?> clazz, long bits) {
+        JfrTraceIdMap map = getTraceIdMap();
+        long id = map.getId(clazz);
+        map.setId(clazz, (id & ~0xff) | (bits & 0xff));
+    }
+
+    public static boolean predicate(Class<?> clazz, long bits) {
+        JfrTraceIdMap map = getTraceIdMap();
+        long id = map.getId(clazz);
+        return (id & bits) != 0;
+    }
+
+    @Uninterruptible(reason = "Called from uninterruptible code.", mayBeInlined = true)
```

Comment:
`mayBeInlined = true` may cause issues for this method. Accessing the `epoch` is a non-volatile read, so when the compiler inlines the method in interruptible code, it can move the read so that the compiled code uses an outdated value even though the value changed at a safepoint.

I would use `@Uninterruptible(reason = "Epoch must not change.")` for all methods that call `thisEpochBit()` in this class.